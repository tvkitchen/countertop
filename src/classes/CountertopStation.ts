import { v4 as uuid } from 'uuid'
import { Kafka } from 'kafkajs'
import { CountertopStationState } from '../types'
import { CountertopStationStateError } from '../errors'
import { CountertopWorker } from './CountertopWorker'
import type { Logger } from '../types'
import type {
	ImplementedApplianceClass,
	ApplianceSettings,
} from './AbstractAppliance'
import type { CountertopTopology } from './CountertopTopology'

interface CountertopStationSettings {
	logger?: Logger;
	kafka: Kafka;
}

export class CountertopStation {
	public readonly settings: CountertopStationSettings

	public readonly id: string

	public readonly ApplianceClass: ImplementedApplianceClass

	public readonly applianceSettings?: ApplianceSettings

	#workers: CountertopWorker[]

	#state: CountertopStationState

	constructor(
		ApplianceClass: ImplementedApplianceClass,
		applianceSettings?: ApplianceSettings,
		settings: CountertopStationSettings = {
			kafka: new Kafka({ brokers: [] }),
		},
	) {
		this.ApplianceClass = ApplianceClass
		this.applianceSettings = applianceSettings
		this.settings = settings
		this.id = `${ApplianceClass.name}::${uuid()}`
		this.#state = CountertopStationState.Stopped
		this.#workers = []
	}

	/**
	 * Replace the station's workers in order to fit a specified topology.
	 *
	 * @param  {CountertopToplogy} topology The topology to use.
	 * @return {CountertopWorker[]}         The new set of workers.
	 */
	public invokeTopology(
		topology: CountertopTopology,
	): CountertopWorker[] {
		this.settings.logger?.debug(`CountertopStation<${this.id}>: invokeTopology()`)
		if (this.state !== CountertopStationState.Stopped) {
			throw new CountertopStationStateError('CountertopStations must be stopped when invoking a topology.')
		}

		const stationStreams = topology.streams.filter(
			(stream) => stream.mouth === this,
		)
		this.#workers = stationStreams.map((stream) => (
			new CountertopWorker(
				this.ApplianceClass,
				this.applianceSettings,
				{
					stream,
					logger: this.settings.logger,
					kafka: this.settings.kafka,
				},
			)
		))
		return this.#workers
	}

	/**
	 * Start the CountertopStation and begin data processing.
	 */
	public async start(): Promise<void> {
		this.settings.logger?.debug(`CountertopStation<${this.id}>: start()`)
		this.#state = CountertopStationState.Starting
		try {
			await this.startAllWorkers()
		} catch (e: unknown) {
			this.#state = CountertopStationState.Errored
			await this.stopAllWorkers()
			throw e
		}
		this.#state = CountertopStationState.Started
	}

	/**
	 * Stop the CountertopStation and halt all data processing.
	 */
	public async stop(): Promise<void> {
		this.settings.logger?.debug(`CountertopStation<${this.id}>: stop()`)
		this.#state = CountertopStationState.Stopping
		await this.stopAllWorkers()
	}

	/**
	 * Get the data types that this station processes.
	 *
	 * This is determined by the Appliance that the station hosts.
	 */
	public getInputTypes() {
		return this.ApplianceClass.getInputTypes(this.applianceSettings)
	}

	/**
	 * Get the data types that this station produces.
	 *
	 * This is determined by the Appliance that the station hosts.
	 */
	public getOutputTypes() {
		return this.ApplianceClass.getOutputTypes(this.applianceSettings)
	}

	public get state() {
		return this.#state
	}

	/**
	 * Registers a listener to the CountertopStation for a given event type.
	 *
	 * Event types are defined in @tvkitchen/base-constants
	 */
	public on(
		eventType: string,
		listener: (...args: unknown[]) => void,
	): this {
		this.#workers.forEach(
			(worker) => worker.on(eventType, listener),
		)
		return this
	}

	private async startAllWorkers(): Promise<void> {
		await Promise.all(
			this.#workers.map(
				async (worker) => worker.start(),
			),
		)
	}

	private async stopAllWorkers(): Promise<void> {
		await Promise.all(
			this.#workers.map(
				async (worker) => worker.stop(),
			),
		)
	}
}
