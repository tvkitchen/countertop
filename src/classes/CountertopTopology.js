import CountertopStream from './CountertopStream'
import {
	getCollectiveOutputTypes,
	getStationsThatConsumeTypes,
	getLongestStreamLength,
	filterStreamsContainingStation,
	generateTributaryMaps,
	getStreamOutputMap,
	getSourceStations,
	identifyObsoleteStreams,
	pruneStreams,
} from '../tools/utils/countertop'

/**
 * CountertopToplogies are a complete set of valid *paths* that data might take across a given
 * array of CountertopStations.
 *
 * The distinction of paths (as opposed to edges / direct steps) is important, as a given payload
 * will have a source, and only payloads with a common source should be considered together in
 * the case of dual-input appliances.
 *
 * In many data processing pipelines, topologies need to be manually specified by the developer
 * configuring the system. In this case, the topologies are automatically generated based on which
 * appliances produce outputs that are valid inputs to others.
 *
 * The logic for generating a CountertopTopology starts in the static `generateStreams` method.
 */
class CountertopTopology {
	stations = []

	streams = []

	/**
	 * Create a CountertopTopology.
	 *
	 * @param  {CountertopStation[]} stations The stations used to define the topology.
	 */
	constructor(stations = []) {
		this.stations = stations
		this.streams = CountertopTopology.generateStreams(this.stations)
	}

	/**
	 * Extend a set of streams by a single station.
	 *
	 * This will only create streams that pull from the same source.
	 * This will generate and return any valid streams that extend a given stream set.
	 * This will ONLY return new streams, which means if no valid streams extensions exist, it will
	 * return an empty array.
	 * This will not create streams that re-visit a given station (no loops).
	 *
	 * @param  {CountertopStream[]}  streams The set of streams to be extended.
	 * @param  {CountertopStation[]} station The station being used to extend.
	 * @return {CountertopStream[]}          The extended streams.
	 */
	static extendStreamsByStation = (streams, station) => {
		const streamOutputMap = getStreamOutputMap(
			filterStreamsContainingStation(streams),
		)
		const tributaryMaps = generateTributaryMaps(
			station,
			streamOutputMap,
		)
		return tributaryMaps.map(
			(tributaryMap) => new CountertopStream(station, tributaryMap),
		)
	}

	/**
	 * Fully extends a set of streams to flow through a set of new stations.
	 *
	 * This will only create streams that pull from the same source.
	 * This will not create streams that re-visit a given station (no loops).
	 * This will create distinct streams for incremental steps from a source (e.g. A->B, and A->B->C).
	 *
	 * @param  {CountertopStream[]}  streams  The base set of streams being extended.
	 * @param  {CountertopStation[]} stations The stations being used to extend the streams.
	 * @return {CountertopStream[]}           The complete set of streams.
	 */
	static generateStreams = (stations, streams = []) => {
		let nextStreams = []
		if (streams.length === 0) {
			nextStreams = CountertopTopology.generateSourceStreams(stations)
		} else {
			const extentionLength = 1 + getLongestStreamLength(streams)
			const outputTypes = getCollectiveOutputTypes(streams)
			const nextStations = getStationsThatConsumeTypes(stations, outputTypes)
			nextStreams = nextStations.flatMap(
				(station) => CountertopTopology.extendStreamsByStation(streams, station),
			)
				// Remove any new streams that aren't long enough
				.filter((stream) => stream.getLength() === extentionLength)
		}
		// Prune any past streams whose tributaries are explicit subsets of new streams
		const obsoleteStreams = identifyObsoleteStreams(streams, nextStreams)
		const prunedStreams = pruneStreams(streams, obsoleteStreams)
		const prunedNextStreams = pruneStreams(nextStreams, obsoleteStreams)

		// If there were any new streams, iterate again
		if (prunedNextStreams.length !== 0) {
			return CountertopTopology.generateStreams(
				stations,
				prunedStreams.concat(prunedNextStreams),
			)
		}
		return prunedStreams
	}

	/**
	 * Generates streams that only include source stations.
	 *
	 * These streams will, by defintiion, be of length 1.
	 *
	 * @param  {CountertopStations[]} stations The stations to be processed.
	 * @return {CountertopStreams[]}           The resulting streams.
	 */
	static generateSourceStreams = (stations) => getSourceStations(stations)
		.map((station) => new CountertopStream(station))
}

export default CountertopTopology
