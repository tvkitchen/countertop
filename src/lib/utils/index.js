// ------------------------------------------------------------------------------------------------
// NB: There are several places in this file that we'd like to use the logger, but unfortunately,
// importing it creates a cyclical dependency (since this file is imported by config.js which is
// imported by logger.js). I've indicated places below where we would like to log, if we ever sort
// out how to safely do so.
// ------------------------------------------------------------------------------------------------

// Given an array of items, returns a new array with only the unique items.
//
// TODO: This could have quite a bit more safety and error-checking built in.
export const uniqueArray = (array) => [...new Set(array)]

// Given two integers, generates all the integers in between.
//
// By default, will include the original two boundary integers. To exclude them, set
// `inclusive: false` in the third (options) argument.
//
// This function does not do type coercion, so boundary arguments must be integers.
export const generateIntegersBetween = (a, b, { inclusive = true } = {}) => {
	// Bail out early if we weren't passed integers
	if (!Number.isInteger(a) || !Number.isInteger(b)) {
		// TODO: Enable the next line once logger-importing is solved:
		// logger.warn('generateIntegersBetween() requires the first two arguments to be integers.')
		return []
	}

	// Bail out early if they're the same number
	if (a === b) {
		return inclusive ? [a] : []
	}

	// Fill that range up! (NB: This is written for clarity, not succinctness.)
	const range = []
	if (inclusive) {
		range.push(a)
	}
	if (a < b) {
		// Ordered lowest to highest
		let current = a + 1
		while (current < b) {
			range.push(current)
			current += 1
		}
	} else {
		// Ordered highest to lowest
		let current = a - 1
		while (current > b) {
			range.push(current)
			current -= 1
		}
	}
	if (inclusive) {
		range.push(b)
	}

	return range
}

// Takes a string range like `"1-4"` and returns an array of boundary integers: `[1,4]`.
//
// If the string contains just a single number (no range), it will return a single-integer array.
//
// Strings containing multiple ranges, e.g. `"1-4-3"`, are officially unsupported, but it tries to
// recover from it by returning the first and last elements of the range. (It does not use min/max
// cleverness to pluck out the highest and lowest numbers in the string, because we support
// reverse-order ranges such as `"6-1"`, and determining order intent from multiple-range strings
// isn't compatible with this.)
//
// Examples:
// - `"1"`   => `[1]`
// - `"1-3"` => `[1,3]`
// - `"3-1"` => `[3,1]`
// - `"1-3-5"` => `[1,5]` // Don't do this!
//
// This method is not broadly useful, but it's still exported so we can test it.
export const splitRangeToIntegers = (range) => {
	let array = []
	if (range.includes('-')) {
		const splitRange = range.split('-')
		// TODO: Enable the next block once logger-importing is solved:
		// if (splitRange.length > 2) {
		// 	logger.warn(`splitRangeToIntegers() was passed a multiple-range string. The result may
		// 		         be unexpected.`)
		// }
		array = [splitRange[0], splitRange.pop()]
	} else {
		array = [range]
	}
	return array.map((i) => parseInt(i, 10))
}

// Converts an integer range shorthand syntax to an array of unique integers.
//
// It receives a string like `"1-5,10-15,19"` and returns all the integers we think that range
// intends to describe. In that case: `[1,2,3,4,5,10,11,12,13,14,15,19]`.
export const expandRangeOfIntegers = (rangeShorthand) => {
	let expandedRange = []
	rangeShorthand.split(',').forEach((segment) => {
		let splitSegments = splitRangeToIntegers(segment)
		if (splitSegments.length === 2) {
			splitSegments = generateIntegersBetween(splitSegments[0], splitSegments[1])
		}
		expandedRange = expandedRange.concat(splitSegments)
	})
	return uniqueArray(expandedRange)
}
