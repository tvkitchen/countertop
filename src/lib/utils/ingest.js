/* eslint-disable import/prefer-default-export */

import 'module-alias/register'

import config from '%src/config'

export const chooseAvailableInternalStream = () => `${config.INTERNAL_STREAM_ADDRESS}:${config.INTERNAL_STREAM_PORT_RANGE[0]}`
