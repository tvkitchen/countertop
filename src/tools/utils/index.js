// Disabling because we intend to have more exports in the future.
/* eslint-disable import/prefer-default-export */

/**
 * Checks whether a method exists in a given object.
 *
 * @param  {String} methodName The name of the method being checked
 * @param  {Object} object     The object in which the method should exist
 * @return {Boolean}           Whether the method is defined and is a function
 */
export const methodExists = (methodName, object) => typeof object[methodName] === 'function'

/**
 * Checks whether two arrays share any items
 * @param  {Array} arr1 The first array being considered.
 * @param  {Array} arr2 The second array being considered.
 * @return {Boolean}    Whether the two arrays share any items.
 */
export const arraysHaveOverlap = (arr1, arr2) => arr1.some(
	(item) => arr2.includes(item),
)

/**
 * Generates a method that can be used to sort a list of objects by
 * the specified attribute.
 * @param  {String}   attribute The attribute to use for the sort.
 * @return {Function}           The sort comparison function.
 */
export const by = (attribute) => (a, b) => ((a[attribute] > b[attribute]) ? 1 : -1)

export * from './sortedArray'
