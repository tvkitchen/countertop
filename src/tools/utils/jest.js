import fs from 'fs'
import path from 'path'
import { IAppliance } from '@tvkitchen/base-interfaces'

/**
 * Loads a file containing JSON from the test's `data` directory and returns the resulting object.
 *
 * @param  {String} fileName The name of the file located in the `data` directory
 */
export const loadTestData = (testDirectory, fileName) => JSON.parse(fs.readFileSync(
	path.join(testDirectory, 'data', fileName),
))

/**
 * Creates a mock appliance class
 *
 * @param  {String[]} options.inputTypes  The mocked input types value.
 * @param  {String[]} options.outputTypes The mocked output types value.
 * @return {Class}                        The resulting mock class.
 */
export const generateMockAppliance = ({
	inputTypes,
	outputTypes,
	getInputTypes = () => inputTypes,
	getOutputTypes = () => outputTypes,
	healthCheck = async () => true,
	start = async () => true,
	stop = async () => true,
}) => {
	class MockAppliance extends IAppliance {
		static getInputTypes = getInputTypes

		static getOutputTypes = getOutputTypes

		on = () => true

		healthCheck = healthCheck

		start = start

		stop = stop
	}
	return MockAppliance
}

/**
 * Normalize a tributary map so that all stations are replaced with a normalized
 * index.
 *
 * @param  {Map}                 streamMap The stream map to normalize.
 * @param  {CountertopStation[]} stations  The stations being normalized.
 * @return {Map}                           The normalized stream map.
 */
export const normalizeTributaryMap = (streamMap, stations) => {
	const normalizedTributaryMap = new Map()
	streamMap.forEach(
		(value, key) => {
			normalizedTributaryMap.set(
				key,
				// https://github.com/eslint/eslint/issues/12473
				// eslint-disable-next-line no-use-before-define
				normalizeStream(value, stations),
			)
		},
	)
	return normalizedTributaryMap
}

/**
 * Normalize an array of tributary maps so that all stations are replaced with a
 * normalized index.
 *
 * @param  {Map[]}               streamMap The stream maps to normalize.
 * @param  {CountertopStation[]} stations  The stations being normalized.
 * @return {Map[]}                         The normalized stream maps.
 */
export const normalizeTributaryMaps = (streamMaps, stations) => streamMaps.map(
	(streamMap) => normalizeTributaryMap(streamMap, stations),
)

/**
 * Normalizes a stream so that all station references are replaced with a normalized
 * index.
 *
 * @param  {CountertopStream}    stream   The stream to normalize.
 * @param  {CountertopStation[]} stations The stations being normalized.
 * @return {CountertopStream}             The normalized stream.
 */
export const normalizeStream = (stream, stations) => ({
	source: stations.findIndex((station) => station.id === stream.getSource().id),
	mouth: stations.findIndex((station) => station.id === stream.getMouth().id),
	tributaryMap: normalizeTributaryMap(stream.getTributaryMap(), stations),
})

/**
 * Normalizes an array of streams so that all station references are replaced with a normalized
 * index.
 *
 * @param  {CountertopStream[]}  stream   The streams to normalize.
 * @param  {CountertopStation[]} stations The stations being normalized.
 * @return {CountertopStream[]}           The normalized streams.
 */
export const normalizeStreams = (streams, stations) => streams.map(
	(stream) => normalizeStream(stream, stations),
)
