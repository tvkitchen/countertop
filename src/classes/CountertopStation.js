import { v4 as uuid } from 'uuid'
import { countertopStates } from '../constants'
import { consoleLogger } from '../tools/loggers'
import { CountertopWorker } from './CountertopWorker'

class CountertopStation {
	logger = null

	kafka = null

	id = null

	Appliance = null

	applianceSettings = null

	workers = []

	state = ''

	/**
	 * Create a CountertopStation
	 *
	 * @param  {Class}  Appliance         The IAppliance class that this station will manage.
	 * @param  {Object} applianceSettings Settings for the appliance.
	 * @param  {Logger} options.logger    A logger with methods for all TV Kitchen logLevels.
	 */
	constructor(
		Appliance,
		applianceSettings = {},
		{
			logger = consoleLogger,
			kafka,
		} = {},
	) {
		if (Appliance === undefined) {
			throw new Error('CountertopStation requires an Appliance.')
		}
		this.Appliance = Appliance
		this.applianceSettings = applianceSettings
		this.logger = logger
		this.kafka = kafka
		this.id = `${Appliance.name}::${uuid()}`
		this.setState(countertopStates.STOPPED)
	}

	/**
	 * Replace the station's workers in order to fit a specified topology.
	 *
	 * @param  {CountertopToplogy} topology The topology to use.
	 * @return {CountertopWorker[]}         The new set of workers.
	 */
	invokeTopology = (topology) => {
		this.logger.trace(`CountertopStation<${this.id}>: invokeTopology()`)
		if (this.getState() !== countertopStates.STOPPED) {
			throw new Error('CountertopStations must be stopped before invoking a topology.')
		}
		const stationStreams = topology.streams.filter((stream) => stream.getMouth() === this)
		this.workers = stationStreams.map((stream) => new CountertopWorker(
			this.Appliance,
			this.applianceSettings,
			{
				stream,
				logger: this.logger,
				kafka: this.kafka,
			},
		))
		return this.workers
	}

	/**
	 * Start the station and begin data processing.
	 *
	 * @return {Boolean} Whether the station successfully started.
	 */
	start = async () => {
		this.logger.trace(`CountertopStation<${this.id}>: start()`)
		this.state = countertopStates.STARTING
		const promises = this.workers.map(async (worker) => worker.start())
		const started = (await Promise.all(promises)).every((result) => result)
		this.setState(started ? countertopStates.STARTED : countertopStates.ERRORED)
		return this.getState() === countertopStates.STARTED
	}

	/**
	 * Stop the countertop and halt all data processing.
	 *
	 * @return {Boolean} Whether the station successfully stopped.
	 */
	stop = async () => {
		this.logger.trace(`CountertopStation<${this.id}>: stop()`)
		this.setState(countertopStates.STOPPING)
		const promises = this.workers.map(async (worker) => worker.stop())
		const stopped = (await Promise.all(promises)).every((result) => result)
		this.setState(stopped ? countertopStates.STOPPED : countertopStates.ERRORED)
		return this.getState() === countertopStates.STOPPED
	}

	/**
	 * Get the data types that this station processes.
	 *
	 * This is determined by the Appliance that the station hosts.
	 *
	 * @return {String[]} The data types that this station processes.
	 */
	getInputTypes = () => this.Appliance.getInputTypes(this.applianceSettings)

	/**
	 * Get the data types that this station produces.
	 *
	 * This is determined by the Appliance that the station hosts.
	 *
	 * @return {String[]} The data types that this station produces.
	 */
	getOutputTypes = () => this.Appliance.getOutputTypes(this.applianceSettings)

	/**
	 * Get the current state of the CountertopStation.
	 *
	 * @return {String} The state of the CountertopStation.
	 */
	getState = () => this.state

	/**
	 * Set the current state of the CountertopStation.
	 *
	 * This is an internal method.
	 */
	setState = (state) => { this.state = state }

	/**
	 * Registers a listener to the CountertopStation for a given event type.
	 *
	 * Event types are defined in @tvkitchen/base-constants
	 *
	 * @param  {String} eventType  The type of event being listened to.
	 * @param  {Function} listener The listener to be registered for that event.
	 * @return {CountertopStation} The CountertopStation instance (to enable chaining).
	 */
	on = (eventType, listener) => {
		this.workers.forEach((worker) => worker.on(eventType, listener))
		return this
	}
}

export default CountertopStation
