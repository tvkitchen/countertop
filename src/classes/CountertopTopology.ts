import {
	getCollectiveOutputTypes,
	getStationsThatConsumeTypes,
	getLongestStreamLength,
	filterStreamsContainingStation,
	generateTributaryMaps,
	getStreamOutputMap,
	getSourceStations,
	identifyObsoleteStreams,
	pruneStreams,
} from '../tools/utils/countertop'
import { CountertopStream } from './CountertopStream'
import { CountertopStation } from './CountertopStation'

/**
 * CountertopToplogies are a complete set of valid *paths* that data might take across a given
 * array of CountertopStations.
 *
 * The distinction of paths (as opposed to edges / direct steps) is important, as a given payload
 * will have a source, and only payloads with a common source should be considered together in
 * the case of dual-input appliances.
 *
 * In many data processing pipelines, topologies need to be manually specified by the developer
 * configuring the system. In this case, the topologies are automatically generated based on which
 * appliances produce outputs that are valid inputs to others.
 *
 * The logic for generating a CountertopTopology starts in the static `generateStreams` method.
 */
export class CountertopTopology {
	stations: CountertopStation[]

	streams: CountertopStream[]

	constructor(stations: CountertopStation[] = []) {
		this.stations = stations
		this.streams = CountertopTopology.generateStreams(this.stations)
	}

	/**
	 * Fully extends a set of streams to flow through a set of new stations.
	 *
	 * This will only create streams that pull from the same source.
	 * This will not create streams that re-visit a given station (no loops).
	 * This will create distinct streams for incremental steps from a source (e.g. A->B, and A->B->C).
	 *
	 * @param  {CountertopStream[]}  streams  The base set of streams being extended.
	 * @param  {CountertopStation[]} stations The stations being used to extend the streams.
	 * @return {CountertopStream[]}           The complete set of streams.
	 */
	public static generateStreams(
		stations: CountertopStation[],
		streams: CountertopStream[] = [],
	): CountertopStream[] {
		let nextStreams = []
		if (streams.length === 0) {
			nextStreams = CountertopTopology.generateSourceStreams(stations)
		} else {
			const extentionLength = 1 + getLongestStreamLength(streams)
			const outputTypes = getCollectiveOutputTypes(streams)
			const nextStations = getStationsThatConsumeTypes(stations, outputTypes)
			nextStreams = nextStations.flatMap(
				// Disable until countertop utilities get types
				// eslint-disable-next-line
				(station) => CountertopTopology.extendStreamsByStation(streams, station),
			)
				// Remove any new streams that aren't long enough
				.filter((stream) => stream.getLength() === extentionLength)
		}
		// Prune any past streams whose tributaries are explicit subsets of new streams
		const obsoleteStreams = identifyObsoleteStreams(streams, nextStreams)
		// Disable until countertop utilities get types
		// eslint-disable-next-line
		const prunedStreams = pruneStreams(streams, obsoleteStreams)
		// Disable until countertop utilities get types
		// eslint-disable-next-line
		const prunedNextStreams = pruneStreams(nextStreams, obsoleteStreams)

		// If there were any new streams, iterate again
		// Disable until countertop utilities get types
		// eslint-disable-next-line
		if (prunedNextStreams.length !== 0) {
			return CountertopTopology.generateStreams(
				stations,
				// Disable until countertop utilities get types
				// eslint-disable-next-line
				prunedStreams.concat(prunedNextStreams),
			)
		}
		// Disable until countertop utilities get types
		// eslint-disable-next-line
		return prunedStreams
	}

	/**
	 * Extend a set of streams by a single station.
	 *
	 * This will only create streams that pull from the same source.
	 * This will generate and return any valid streams that extend a given stream set.
	 * This will ONLY return new streams, which means if no valid streams extensions exist, it will
	 * return an empty array.
	 * This will not create streams that re-visit a given station (no loops).
	 *
	 * @param  {CountertopStream[]}  streams The set of streams to be extended.
	 * @param  {CountertopStation[]} station The station being used to extend.
	 * @return {CountertopStream[]}          The extended streams.
	 */
	private static extendStreamsByStation(
		streams: CountertopStream[],
		station: CountertopStation,
	): CountertopStream[] {
		const streamOutputMap = getStreamOutputMap(
			filterStreamsContainingStation(streams, station),
		)
		const tributaryMaps = generateTributaryMaps(
			station,
			streamOutputMap,
		)
		return tributaryMaps.map(
			// Disable until countertop utilities get types
			// eslint-disable-next-line
			(tributaryMap) => new CountertopStream(station, tributaryMap),
		)
	}

	/**
	 * Generates streams that only include source stations.
	 *
	 * These streams will, by defintiion, be of length 1.
	 *
	 * @param  {CountertopStations[]} stations The stations to be processed.
	 * @return {CountertopStreams[]}           The resulting streams.
	 */
	private static generateSourceStreams(stations: CountertopStation[]) {
		return getSourceStations(stations)
			// Disable until countertop utilities get types
			// eslint-disable-next-line
			.map((station) => new CountertopStream(station))
	}
}
