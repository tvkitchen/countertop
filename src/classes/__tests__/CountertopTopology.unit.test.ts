import { CountertopTopology } from '../CountertopTopology'
import { CountertopStation } from '../CountertopStation'
import { generateMockAppliance } from '../../tools/test'
import { normalizeStreams } from '../../tools/utils/jest'

describe('CountertopTopology #unit', () => {
	describe('constructor', () => {
		it('Should generate no streams if provided no stations', () => {
			const topology = new CountertopTopology()
			expect(topology.streams).toEqual([])
		})
	})

	describe('generateStreams', () => {
		it('Should generate the correct streams', () => {
			const ApplianceA = generateMockAppliance({
				inputTypes: [],
				outputTypes: ['foo'],
			})
			const ApplianceB = generateMockAppliance({
				inputTypes: ['foo'],
				outputTypes: ['bar'],
			})
			const stationA = new CountertopStation(ApplianceA)
			const stationB = new CountertopStation(ApplianceB)
			const stations = [stationA, stationB]
			const streams = CountertopTopology.generateStreams(stations)
			expect(normalizeStreams(streams, stations)).toMatchSnapshot()
		})
		it('Should not generate streams for stations with no inputs', () => {
			const ApplianceA = generateMockAppliance({
				inputTypes: [],
				outputTypes: ['foo'],
			})
			const ApplianceB = generateMockAppliance({
				inputTypes: ['baz'],
				outputTypes: ['bar'],
			})
			const stationA = new CountertopStation(ApplianceA)
			const stationB = new CountertopStation(ApplianceB)
			const stations = [stationA, stationB]
			const streams = CountertopTopology.generateStreams(stations)
			expect(normalizeStreams(streams, stations)).toMatchSnapshot()
		})
		it('Should not generate redundant partial streams for stations with multiple inputs', () => {
			const ApplianceA = generateMockAppliance({
				inputTypes: [],
				outputTypes: ['foo'],
			})
			const ApplianceB = generateMockAppliance({
				inputTypes: ['foo'],
				outputTypes: ['bar'],
			})
			const ApplianceC = generateMockAppliance({
				inputTypes: ['foo', 'bar'],
				outputTypes: ['baz'],
			})
			const stationA = new CountertopStation(ApplianceA)
			const stationB = new CountertopStation(ApplianceB)
			const stationC = new CountertopStation(ApplianceC)
			const stations = [stationA, stationB, stationC]
			const streams = CountertopTopology.generateStreams(stations)
			expect(normalizeStreams(streams, stations)).toMatchSnapshot()
		})
		it('Should maintain complete toplogies after redundant partial streams are removed', () => {
			const ApplianceA = generateMockAppliance({
				inputTypes: [],
				outputTypes: ['foo'],
			})
			const ApplianceB = generateMockAppliance({
				inputTypes: ['foo'],
				outputTypes: ['bar'],
			})
			const ApplianceC = generateMockAppliance({
				inputTypes: ['foo', 'bar'],
				outputTypes: ['baz'],
			})
			const ApplianceD = generateMockAppliance({
				inputTypes: ['baz'],
				outputTypes: ['bop'],
			})
			const stationA = new CountertopStation(ApplianceA)
			const stationB = new CountertopStation(ApplianceB)
			const stationC = new CountertopStation(ApplianceC)
			const stationD = new CountertopStation(ApplianceD)
			const stations = [stationA, stationB, stationC, stationD]
			const streams = CountertopTopology.generateStreams(stations)
			expect(normalizeStreams(streams, stations)).toMatchSnapshot()
		})
	})
})
