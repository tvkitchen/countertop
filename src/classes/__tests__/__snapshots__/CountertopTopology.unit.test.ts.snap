// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountertopTopology #unit generateStreams Should generate the correct streams 1`] = `
Array [
  Object {
    "mouth": 0,
    "source": 0,
    "tributaryMap": Map {},
  },
  Object {
    "mouth": 1,
    "source": 0,
    "tributaryMap": Map {
      "foo" => Object {
        "mouth": 0,
        "source": 0,
        "tributaryMap": Map {},
      },
    },
  },
]
`;

exports[`CountertopTopology #unit generateStreams Should maintain complete toplogies after redundant partial streams are removed 1`] = `
Array [
  Object {
    "mouth": 0,
    "source": 0,
    "tributaryMap": Map {},
  },
  Object {
    "mouth": 1,
    "source": 0,
    "tributaryMap": Map {
      "foo" => Object {
        "mouth": 0,
        "source": 0,
        "tributaryMap": Map {},
      },
    },
  },
  Object {
    "mouth": 2,
    "source": 0,
    "tributaryMap": Map {
      "foo" => Object {
        "mouth": 0,
        "source": 0,
        "tributaryMap": Map {},
      },
      "bar" => Object {
        "mouth": 1,
        "source": 0,
        "tributaryMap": Map {
          "foo" => Object {
            "mouth": 0,
            "source": 0,
            "tributaryMap": Map {},
          },
        },
      },
    },
  },
  Object {
    "mouth": 3,
    "source": 0,
    "tributaryMap": Map {
      "baz" => Object {
        "mouth": 2,
        "source": 0,
        "tributaryMap": Map {
          "foo" => Object {
            "mouth": 0,
            "source": 0,
            "tributaryMap": Map {},
          },
          "bar" => Object {
            "mouth": 1,
            "source": 0,
            "tributaryMap": Map {
              "foo" => Object {
                "mouth": 0,
                "source": 0,
                "tributaryMap": Map {},
              },
            },
          },
        },
      },
    },
  },
]
`;

exports[`CountertopTopology #unit generateStreams Should not generate redundant partial streams for stations with multiple inputs 1`] = `
Array [
  Object {
    "mouth": 0,
    "source": 0,
    "tributaryMap": Map {},
  },
  Object {
    "mouth": 1,
    "source": 0,
    "tributaryMap": Map {
      "foo" => Object {
        "mouth": 0,
        "source": 0,
        "tributaryMap": Map {},
      },
    },
  },
  Object {
    "mouth": 2,
    "source": 0,
    "tributaryMap": Map {
      "foo" => Object {
        "mouth": 0,
        "source": 0,
        "tributaryMap": Map {},
      },
      "bar" => Object {
        "mouth": 1,
        "source": 0,
        "tributaryMap": Map {
          "foo" => Object {
            "mouth": 0,
            "source": 0,
            "tributaryMap": Map {},
          },
        },
      },
    },
  },
]
`;

exports[`CountertopTopology #unit generateStreams Should not generate streams for stations with no inputs 1`] = `
Array [
  Object {
    "mouth": 0,
    "source": 0,
    "tributaryMap": Map {},
  },
]
`;
